src/MacroExpander.js

```diff
@@ -25,18 +28,87 @@ MacroExpander.prototype.nextToken = function() {
         if (!(name.charAt(0) === "\\" && this.macros.hasOwnProperty(name))) {
             return topToken;
         }
+        var tok;
         var expansion = this.macros[name];
         if (typeof expansion === "string") {
+            var numArgs = 0;
+            if (expansion.indexOf("#") !== -1) {
+                var stripped = expansion.replace(/##/g, "");
+                while (stripped.indexOf("#" + (numArgs + 1)) !== -1) {
+                    ++numArgs;
+                }
+            }
             var bodyLexer = new Lexer(expansion);
             expansion = [];
-            var tok = bodyLexer.lex();
+            tok = bodyLexer.lex();
             while (tok.text !== "EOF") {
                 expansion.push(tok);
                 tok = bodyLexer.lex();
             }
             expansion.reverse(); // to fit in with stack using push and pop
+            expansion.numArgs = numArgs;
             this.macros[name] = expansion;
         }
+        if (expansion.numArgs) {
```

Comment:
Yes, this is but the first step.

But macros inside the arguments already work. Try `\underset-{\overset+/}` if you want to. That's because expansion gets triggered when a token is removed from the stack, and works by putting the expanded form on the stack. So the arguments end up on the stack, from where they will get expanded once the parser gets there. Which I think is what LaTeX does, too.

Macros with unbalanced `}` in them are somewhat tricky. As far as I know, you can't really define them in TeX either, at least not using that notation. That's what `\bgroup` and `\egroup` are for. Come to think of it, it should be perfectly possible to use our built-in macro definition facility to define `\bgroup` and `\egroup`. The requirement to have balanced brackets is in the *argument* handling, and it's in the Parser down the line. Neither of these should matter. Will give that a try. By the way, do we care about [the distinction between `\bgroup` and `\begingroup`](http://tex.stackexchange.com/q/1930/16923)?

The case of fixed strings in the macro definition is something @edemaine also already mentioned. I think we should do that eventually, but right now I don't even know how TeX does that, exactly. Will need to figure that out first, then I can see about implementing it. Might add a comment until then.